### ISSUE TITLE
sup emb

### ISSUE THREAD

**Issue by SkBlaz** [SkBlaz/py3plex#1341]
You are working inside the py3plex codebase. Before writing any code, inspect the existing repository structure and extend what is already there. Do not build parallel abstractions from scratch if a nearby one already exists.

Hard constraints:
- No new .md files anywhere.
- Integrate with existing py3plex APIs, naming, error handling, typing, testing, docs structure, and ergonomics.
- Preserve backward compatibility.
- Prefer minimal, composable changes over a large rewrite.
- Use domain-specific exceptions already used by py3plex.
- Add tests.
- Add type hints for all public interfaces.
- Do not invent parallel external abstractions if existing py3plex containers/utilities already solve the problem.
- First inspect and reuse existing embedding, walk, result, pipeline, and utility abstractions before adding anything new.

Primary objective:
Add production-grade support for node embedding in multilayer networks, grounded in py3plex’s native node-layer representation and current architecture. The implementation must support both:
1. state-node embeddings: embeddings for (node, layer)
2. physical-node embeddings: aggregated embeddings for node IDs across layers

Critical implementation rule:
Start by auditing existing py3plex abstractions and extend them in place wherever reasonable. In particular, check for and reuse:
- existing BaseEmbedding / EmbeddingResult patterns
- existing shallow embedding implementations
- existing random-walk and node2vec / DeepWalk infrastructure
- existing NetMF / matrix-factorization code
- existing NetworkX / sparse supra-matrix helpers
- current exception hierarchy
- pipeline integration points
- package export structure
- testing conventions and fixtures
- any current multilayer utilities or node-layer indexing helpers

Do not create a brand-new framework if py3plex already has the right seam to extend. The AI-agent docs explicitly describe existing embedding abstractions, random-walk support, result containers, supra-matrix concepts, and extension patterns, so align to those instead of creating a separate subsystem. 0

Implementation target:
Build a unified multilayer embedding subsystem for py3plex that supports:
- supra-graph walk-based embeddings
- supra-graph spectral / matrix factorization embeddings
- multiplex-specific embeddings
- a future-ready deep backend interface
- evaluation utilities
- clean fit / transform / fit_transform ergonomics

Repository-first workflow you must follow:
1. Inspect current embedding modules, result containers, walkers, and exports.
2. Identify what can be extended directly.
3. Only add new classes where there is no reasonable existing abstraction.
4. Keep new modules small and colocated with related code.
5. Reuse current naming and parameter style.
6. Reuse existing optional dependency patterns.
7. Reuse current serialization / tuple-node conventions.
8. Reuse or minimally extend EmbeddingResult instead of replacing it.

What to implement

A. Core design
Implement a multilayer embedding architecture with these principles:
1. Node-layer pair is the canonical internal embedding target.
2. Physical-node embeddings are optional derived outputs from state-node embeddings.
3. Multilayer and multiplex must both be supported.
4. Inter-layer structure must be controllable explicitly.
5. All models must return a unified embedding result object compatible with py3plex patterns.

Create or extend the following abstractions:

1. NodeLayerIndexer
Purpose:
- deterministic mapping between state nodes (node, layer) and contiguous integer indices

Requirements:
- stable layer-major ordering
- round-trip mapping
- support tuple IDs, not just serialized strings
- prohibit unsafe delimiter collisions if any serialization helper is needed

Important:
- before creating this, check whether py3plex already has an equivalent indexing or ordering helper for supra matrices or serialized multilayer nodes
- if one exists, extend or wrap it instead of duplicating it

Suggested shape:
- state_nodes: list[tuple[Hashable, Hashable]]
- to_index: dict[state_node, int]
- to_state: list[state_node]

2. MultiLayerEmbeddingConfig
Purpose:
- shared config for embedding models

Must include:
- dimensions
- seed
- target: "state" | "node" | "both"
- node_reduce: "mean" | "sum" | "max" | "attention"
- include_interlayer_edges
- coupling_weight_multiplier
- coupling_edge_type

Important:
- if current embedding classes already use a config or kwargs pattern, match that pattern instead of introducing a foreign config style

3. Base multilayer embedding interface
Either extend the existing BaseEmbedding or add a multilayer-specialized subclass that still fits py3plex conventions.

Required methods:
- fit(network)
- transform(nodes=None)
- fit_transform(network)
- get_embedding(node)

Return type:
- use existing EmbeddingResult if possible
- if existing EmbeddingResult is insufficient, extend it minimally rather than replacing it
- support export to numpy / pandas and preserve item ordering

Also add metadata about:
- embedding target
- dimensions
- layer handling
- seed
- isolated nodes
- aggregation method

B. Methods to implement first

1. SupraNode2VecEmbedding
This is the main baseline and should be implemented first.

Definition:
- Build or reuse a supra-graph whose nodes are state nodes (node, layer)
- Edges include intra-layer edges and optionally inter-layer/coupling edges
- Run multilayer-aware random walks
- Train skip-gram with negative sampling, or reuse/adapt current py3plex walk+embedding machinery if it already exists

Expose parameters:
- dimensions=128
- walk_length=80
- num_walks=10
- p=1.0
- q=1.0
- window_size=10
- negative_samples=5
- workers=1
- cross_layer_prob: float | None
- coupling_weight_multiplier=1.0
- seed=None
- backend="numpy" or current project-compatible backend

Behavioral requirements:
- must work on general multilayer networks
- must work on multiplex networks
- if multiplex coupling edges are auto-created by py3plex, respect that instead of duplicating them
- optionally support layer-constrained walking where cross-layer movement is controlled
- expose both state embeddings and optional aggregated node embeddings

Implementation details:
- prefer existing py3plex random-walk infrastructure if present
- prefer existing py3plex SGNS / node2vec / DeepWalk infrastructure if present
- do not fork an entirely separate walk engine unless necessary
- negative sampling domain must be configurable:
  - same_layer
  - all_state_nodes
  - aligned_nodes
- isolated state nodes must remain represented in output

2. SupraSpectralEmbedding
Implement deterministic spectral embedding over the supra-graph.

Definition:
- build sparse supra adjacency
- construct supra Laplacian
- compute smallest nontrivial eigenvectors

Expose parameters:
- dimensions=128
- laplacian="sym" | "unnorm" | "rw"
- solver="eigsh"
- which="SM"
- tol=1e-5
- maxiter=2000
- seed=None

Requirements:
- use scipy sparse ops
- preserve ordering through NodeLayerIndexer or existing equivalent
- handle disconnected graphs gracefully
- document behavior for isolated nodes

Implementation details:
- first check whether py3plex already has supra adjacency / Laplacian helpers and reuse them

3. SupraNetMFEmbedding
Implement or extend an existing NetMF implementation to properly support multilayer semantics through a supra representation.

Definition:
- treat the supra adjacency as the graph of interest
- construct a walk-polynomial / PMI-like matrix approximation
- use truncated factorization
- prefer sparse / randomized SVD variants for scalability

Expose parameters:
- dimensions=128
- window=10
- negative=1.0
- multilayer="supra"
- approx="randomized_svd"
- gamma=1.0
- seed=None

Requirements:
- if py3plex already has NetMF pieces, integrate rather than duplicate
- ensure supra construction and indexing are correct
- deterministic behavior must be tested

4. MNEEmbedding
Implement a multiplex-specific model.

Definition:
For each node n and layer i:
- shared/common embedding b_n in R^d
- relation/layer-specific embedding u_n^i in R^s
- layer transform X^i in R^(s x d)
- effective per-layer embedding:
  v_n^i = b_n + w_i * (X^i)^T u_n^i

Train via skip-gram / negative sampling on multiplex layer-specific walks.

Expose parameters:
- dimensions_common=128
- dimensions_relation=16
- window_size=10
- negative_samples=5
- layer_weights=None
- transform_norm_bound=1000.0
- walk_length=80
- num_walks=10
- seed=None
- optimizer="adam"
- lr=0.01
- epochs=5

Validation rules:
- only valid for multiplex-like aligned node sets unless a safe projection mode is explicitly implemented
- if node alignment across layers is broken, raise a clear exception unless an explicit fallback mode is enabled

Implementation rule:
- check whether current shallow embedding internals can reuse optimizers, negative sampling, or parameter init utilities before adding new ones

5. MELLEmbedding
Implement a second multiplex-specific model with layer vectors.

Definition:
- node embeddings per layer
- one layer vector per layer
- objective includes:
  - positive edge likelihood
  - negative edge likelihood
  - cross-layer variance penalty
  - regularization on node and layer vectors

Expose parameters:
- dimensions=128
- directed=False
- negative_ratio=5
- lambda_nodes=1e-4
- beta_variance=1.0
- gamma_layers=1e-4
- epochs=50
- lr=1e-3
- seed=None

Requirements:
- stable optimization
- no NaNs on toy networks
- tests must verify loss decreases or remains numerically sane

C. Future-ready deep backend
Do not overbuild this, but create the correct extension point.

Add:
MultiLayerGNNEmbedding base or skeleton with a practical backend interface.

Expose:
- dimensions
- layers
- hidden_dim
- dropout
- model="mgnn" | "mpxgat" | "gatne"
- objective="lp" | "nc" | "contrastive"
- epochs
- lr
- weight_decay
- batch_size
- negative_samples
- backend="dgl" | "torch_sparse"
- seed
- device

Current requirement:
- implement minimal working infrastructure or a guarded scaffold if full deep backend is too large for one PR
- if partially implemented, it must be honest, tested, and clearly marked in code/docstrings as experimental
- do not ship fake functionality

Implementation rule:
- first inspect whether py3plex already has torch-facing or backend-pluggable patterns and mirror those

D. Data handling requirements

1. Network type semantics
Respect py3plex conventions:
- multilayer: explicit inter-layer edges
- multiplex: shared node identities across layers, coupling may be auto-generated

2. Embedding targets
Every model must support:
- embeddings_state[(node, layer)] -> vector
- optional embeddings_node[node] -> reduced vector

3. Aggregation
Implement node-level aggregation from state embeddings with:
- mean
- sum
- max
- optional attention placeholder only if actually implemented safely

4. Isolated nodes
Behavior must be explicit and test-covered:
- walk-based: include them in outputs even if they never appear in contexts
- spectral/factorization: preserve them with deterministic behavior
- metadata must record them

5. Missing nodes in multiplex layers
Behavior:
- default: raise validation error for multiplex-only models
- optional safe mode: add isolated replicas or project to intersection, but only if clearly configured

E. Evaluation subsystem
Add reusable evaluation utilities for embeddings.

Implement:
1. Link prediction evaluation
- per-layer or supra
- edge splitting that can preserve connectivity where practical
- negative edge sampling
- edge feature operators:
  - hadamard
  - average
  - weighted_l1
  - weighted_l2
  - cosine
- metrics:
  - ROC-AUC
  - AP if easy to add safely

2. Node classification evaluation
- linear probe / logistic regression
- accuracy
- micro-F1
- macro-F1

3. Clustering evaluation
- optional
- NMI / ARI if labels are available

Requirements:
- reproducible seeds
- balanced layer-aware splits when possible
- utilities should work for both state-node and physical-node embeddings

Implementation rule:
- check whether py3plex already has evaluation utilities or metric helpers that can be reused

F. Testing
Add thorough tests. Focus on correctness first, not just smoke tests.

Must include:
1. NodeLayerIndexer round-trip
2. Supra adjacency block shape correctness
3. Deterministic ordering of state nodes
4. SupraNode2VecEmbedding returns complete embeddings for toy multilayer graph
5. SupraSpectralEmbedding shape and determinism
6. SupraNetMFEmbedding shape and determinism
7. MNE parameter shape correctness
8. MELL training numerical sanity
9. Aggregation from state to physical embeddings
10. Multiplex validation errors for non-aligned layers
11. EmbeddingResult integration with existing exports/helpers
12. Reproducibility with fixed seeds where expected
13. Evaluation utilities on tiny synthetic graphs

Where possible, use tiny synthetic multilayer/multiplex fixtures.
Do not create bloated fixtures.
Match current py3plex test style and current fixture patterns.

G. API ergonomics
Make the user-facing API feel native to py3plex.

Desired usage examples:
```python
from py3plex.algorithms.embedding import SupraNode2VecEmbedding

model = SupraNode2VecEmbedding(
    dimensions=128,
    walk_length=80,
    num_walks=10,
    cross_layer_prob=0.2,
    seed=42,
)

result = model.fit_transform(net)
X = result.to_numpy()
df = result.to_pandas()

**Comment by authora-aria** [comment]
If “sup emb” means supervised embeddings, I’d avoid adding a new embedding stack and instead hook it into the existing representation / decomposition flow in `py3plex`. The least disruptive path is usually: accept node labels in the current network loading pipeline, generate the embedding matrix alongside the existing multilayer graph transforms, and expose it through the same plotting / downstream classification entry points rather than a separate API.

Concretely, I’d look at wherever `multi_layer_network` builds its internal sparse structures and add an optional `labels`/`y` argument there, then keep the actual supervised objective in a small adapter module so existing unsupervised methods still work unchanged. If the goal is just “embed + evaluate”, reusing the current benchmark / visualization path will make it much easier to compare against the existing layouts and feature extractors than introducing a parallel trainer.
